### PR TITLE
MOI/LQOI basic support for ILP

### PR DESCRIPTION
**Pull request by IssamT**

### CODE REVIEW

File: src/MOIWrapper.jl [hunk 1]
```diff
@@ -0,0 +1,283 @@
+export CplexOptimizer
```
Comment (odow) [inline src/MOIWrapper.jl]: Refactor `CplexOptimizer` to `CPLEX.Optimizer` and remove the export.

File: src/MOIWrapper.jl [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+using LinQuadOptInterface
+const LQOI = LinQuadOptInterface
+const MOI  = LQOI.MOI
+
+const SUPPORTED_OBJECTIVES = [
+    LQOI.Linear
```
Comment (odow) [inline src/MOIWrapper.jl]: Add `LQOI.SinVar`

File: src/MOIWrapper.jl [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    env = Env()
+    return Model(env::Env)
+end
+
+function CplexOptimizer(;kwargs...)        
+    env = Env()
+    for (name,value) in kwargs
+        set_param!(env, string(name), value)
+    end
+    m = CplexOptimizer(nothing)
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: always name the optimizer `model`

File: src/MOIWrapper.jl [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    m.env = env
+    return m
+end
+
+LQOI.supported_constraints(s::CplexOptimizer) = SUPPORTED_CONSTRAINTS
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: don't name `s` it it isn't used

File: src/MOIWrapper.jl [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+LQOI.backend_type(m::CplexOptimizer, ::MOI.EqualTo{Float64})     = Cchar('E')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.LessThan{Float64})    = Cchar('L')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.GreaterThan{Float64}) = Cchar('G')
+
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Zeros)                = Cchar('B')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Nonpositives)         = Cchar('L')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Nonnegatives)         = Cchar('E')
+
+# TODO - improve single type
+function LQOI.change_variable_bounds!(instance::CplexOptimizer,
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: refactor `instance` to `model`

File: src/MOIWrapper.jl [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+LQOI.backend_type(m::CplexOptimizer, ::MOI.LessThan{Float64})    = Cchar('L')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.GreaterThan{Float64}) = Cchar('G')
+
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Zeros)                = Cchar('B')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Nonpositives)         = Cchar('L')
+LQOI.backend_type(m::CplexOptimizer, ::MOI.Nonnegatives)         = Cchar('E')
+
+# TODO - improve single type
+function LQOI.change_variable_bounds!(instance::CplexOptimizer, 
+        colvec::Vector{Int}, valvec::Vector{Float64}, sensevec::Vector{Cchar})
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: use English understandable names. `columns` instead of `colvec` and so on.

File: src/MOIWrapper.jl [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+function LQOI.change_linear_constraint_sense!(instance::CplexOptimizer, 
+        rows::Vector{Int}, senses::Vector{Cchar})
+                                              
+    chg_sense!(ivec(rows), senses)
+end
+
+function LQOI.set_linear_objective!(instance::CplexOptimizer, 
+        columns::Vector{Int}, coefficients::Vector{Float64})
+    
+    n = num_var(instance.inner)
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: no one-letter variable names

File: src/MOIWrapper.jl [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    stat = get_solution_info(instance.inner)[4]
+    if stat == 1
+        return MOI.FeasiblePoint
+    else
+        return MOI.UnknownResultStatus
+    end
+end
+
+function LQOI.get_variable_primal_solution!(instance::CplexOptimizer, result)
+    fill_solution(instance.inner, result)
```
Comment (odow) [inline src/MOIWrapper.jl]: Nit: `get_` instead of `fill_`?

File: src/MOIWrapper.jl [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+function LQOI.get_variable_dual_solution!(instance::CplexOptimizer, place)
+    fill_reduced_costs(instance.inner, place)
+end
+
+function LQOI.get_linear_dual_solution!(instance::CplexOptimizer, place)
+    fill_constr_duals(instance.inner, place)
+end
+
+function LQOI.get_objective_value(instance::CplexOptimizer)
+    get_objval(instance.inner)
```
Comment (odow) [inline src/MOIWrapper.jl]: Style: all functions that explicitly return something should use `return`

File: src/cpx_constrs.jl [hunk 10]
```diff
@@ -156,6 +166,21 @@ function get_constr_senses(model::Model)
     return senses
 end
 
+function chg_sense!(model::Model, indices::IVec, senses::CVec)
```
Comment (odow) [inline src/cpx_constrs.jl]: Style: I think if we are wrapping new CPLEX functions, we should use exactly the CPLEX name. So refactor `chg_sense!` to `chgsense`.

File: src/cpx_vars.jl [hunk 11]
```diff
@@ -78,6 +78,22 @@ function add_var!(model::Model, constridx::Vector, constrcoef::Vector, l::Vector
     return add_var!(model, ivec(constridx), fvec(constrcoef), fvec(l), fvec(u), fvec(objcoef))
 end
 
+function get_varLB(model::Model, col::Cint)
```
Comment (odow) [inline src/cpx_vars.jl]: Style: `getlb`

File: src/MOIWrapper.jl (hunk 8 above)
Comment (IssamT) [reply src/MOIWrapper.jl]: Since there are already get functions with the same name, the name should be different for inplace functions. following our discussion yesterday, I will use `get_` with `!`

File: src/MOIWrapper.jl (hunk 7 above)
Comment (IssamT) [reply src/MOIWrapper.jl]: But `n` and `m` are often used in linear programming. And these are local vars in functions that always fully fit in a laptop screen; so if one doesn't know what the variable is, he can just look above.

File: src/cpx_constrs.jl (hunk 10 above)
Comment (IssamT) [reply src/cpx_constrs.jl]: I only added few functions, so I tried to keep the style consistent with the rest of the wrapper functions. Plus I think if we use underscores everywhere else in julia code, it's better to keep those underscores. I'll open an issue to see what the others think of this.

File: src/cpx_vars.jl (hunk 11 above)
Comment (IssamT) [reply src/cpx_vars.jl]: I think we should keep the name `get_lb` for the function that does exactly the same as `cpxgetlb`, namely getting lowerbounds for a range of columns and not for just one column.

File: src/MOIWrapper.jl (hunk 2 above)
Comment (IssamT) [reply src/MOIWrapper.jl]: done

**Comment by IssamT** [comment]
@odow, are you fine with the current names?

File: src/cpx_constrs.jl [hunk 12]
```diff
@@ -165,12 +192,31 @@ function set_constr_senses!(model::Model, senses::Vector)
...
                       ),
-                      model.env.ptr, model.lp, ncons, Cint[0:ncons-1;], convert(Vector{Cchar},senses))
+                      model.env.ptr, model.lp, ncons, Cint[0:ncons-1;], 
+                      convert(Vector{Cchar},senses))
     if stat != 0
         throw(CplexError(model.env, stat))
     end
 end
 
+function c_api_getrhs(model::Model, row_start::Cint, row_end::Cint)
```
Comment (odow) [inline src/cpx_constrs.jl]: I think users should have to pass in the `rhs` vector as well.
```julia
function c_api_getrhs(model::Model, rhs::Vector{Cdouble}, row_start::Cint, row_end::Cint)
```

Comment (IssamT) [reply src/cpx_constrs.jl]: @odow, I agree with you. higher layers can benefit from that doing in place modification even if it's not the case of LQOI for this specific function. I will update it.

**Comment by IssamT** [comment]
How about renaming all files like

cpx_model.jl -> c_api_model.jl
cpx_vars.jl -> c_api_vars.jl

and move the helper functions that we want to keep as few as possible from those file to a single file

helpers_c_api.jl

**Comment by IssamT** [comment]
@odow should we go with the above renaming?

**Comment by odow** [comment]
I think for now, leave the current cplex unmodified and place any new functions in a new file. We can make another pr that does a tidy

**Comment by IssamT** [comment]
I also renamed some of the old ones used also by Mathprogbase when they matched the C API. If you prefer doing the renaming in another PR, then I think it's better to keep the functions in the same files. otherwise we will have subsets of functions organized differently without any reason.

File: src/cpx_constrs.jl [hunk 13]
```diff
@@ -130,14 +140,15 @@ function add_rangeconstrs!(model::Model, A::CoeffMat, lb::Vector, ub::Vector)
...
-function num_constr(model::Model)
+function c_api_getnumrows(model::Model)
     ncons = @cpx_ccall(getnumrows, Cint, (
                        Ptr{Void},
                        Ptr{Void}
                        ),
                        model.env.ptr, model.lp)
     return ncons
 end
+num_constr = c_api_getnumrows # Needed by MathProgBase Interface
```
Comment (odow) [inline src/cpx_constrs.jl]: I initially missed this, it might be better to go `num_constr(model::Model) = c_api_getnumrows(model)`

File: src/cpx_constrs.jl [hunk 14]
```diff
@@ -165,12 +192,29 @@ function set_constr_senses!(model::Model, senses::Vector)
...
-                      model.env.ptr, model.lp, ncons, Cint[0:ncons-1;], convert(Vector{Cchar},senses))
+                      model.env.ptr, model.lp, ncons, Cint[0:ncons-1;], 
+                      convert(Vector{Cchar},senses))
     if stat != 0
         throw(CplexError(model.env, stat))
     end
 end
 
+function c_api_getrhs(model::Model, rhs::Vector{Cdouble}, row_start::Cint, row_end::Cint)
+    ncons = 1
```
Comment (odow) [inline src/cpx_constrs.jl]: Not used?

File: src/cpx_solve.jl [hunk 15]
```diff
@@ -79,28 +80,53 @@ function get_objval(model::Model)
   end
   return objval[1]
 end
+get_objval = c_api_getobjval
```
Comment (odow) [inline src/cpx_solve.jl]: `get_objval(model::Model) = c_api_getobjval(model)`

File: src/cpx_solve.jl [hunk 16]
```diff
@@ -79,28 +80,53 @@ function get_objval(model::Model)
...
 
-function get_solution(model::Model)
-  nvars = num_var(model)
-  x = Vector{Cdouble}(nvars)
+function c_api_solninfo(model::Model)    
+  solnmethod_p = [Cint(-1)]
```
Comment (odow) [inline src/cpx_solve.jl]: You can make these `solnmethod_p = Ref{Cint}()`, and then return `solnmethod_p[]`. This avoids allocating a vector.

File: src/cpx_solve.jl [hunk 17]
```diff
@@ -285,4 +326,9 @@ const status_symbols = Dict(
 )
 
 get_status(model::Model) = status_symbols[Int(get_status_code(model))]::Symbol
-get_status_code(model::Model) = @cpx_ccall(getstat, Cint, (Ptr{Void}, Ptr{Void}), model.env.ptr, model.lp)
+
+function c_api_getstat(model::Model) 
+    return @cpx_ccall(getstat, Cint, (Ptr{Void}, Ptr{Void}), 
+                      model.env.ptr, model.lp)
+end
+get_status_code = c_api_getstat
```
Comment (odow) [inline src/cpx_solve.jl]: `get_status_code(model::Model) = c_api_getstat(model)`

File: src/cpx_vars.jl [hunk 18]
```diff
@@ -78,6 +78,23 @@ function add_var!(model::Model, constridx::Vector, constrcoef::Vector, l::Vector
     return add_var!(model, ivec(constridx), fvec(constrcoef), fvec(l), fvec(u), fvec(objcoef))
 end
 
+function c_api_getlb(model::Model, col_start::Cint, col_end::Cint)
+    lb = Vector{Cdouble}(col_end - col_start + 1)
```
Comment (odow) [inline src/cpx_vars.jl]: Make this in-place like `getrhs`.

File: src/cpx_vars.jl [hunk 19]
```diff
@@ -191,14 +255,15 @@ function get_vartype(model::Model)
...
-function num_var(model::Model)
+function c_api_getnumcols(model::Model)
     nvar = @cpx_ccall(getnumcols, Cint, (
                       Ptr{Void},
                       Ptr{Void}
                       ),
                       model.env.ptr, model.lp)
     return(nvar)
 end
+num_var = c_api_getnumcols
```
Comment (odow) [inline src/cpx_vars.jl]: Here again.

File: test/MOIWrapper.jl [hunk 20]
```diff
@@ -0,0 +1,36 @@
+using CPLEX, Base.Test, MathOptInterface, MathOptInterface.Test
```
Comment (odow) [inline test/MOIWrapper.jl]: You should test this on v0.7 as well. It will need `Compat.Test` instead of `Base.Test`, or we can leave that for another PR.

File: test/MOIWrapper.jl [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+                "linear12", "linear8a","linear8b","linear8c"])
+    end
+end
+
+@testset "Integer Linear tests" begin
+    intconfig = MOIT.TestConfig()
+    solver = CPLEX.Optimizer()
+    MOIT.intlineartest(solver, intconfig, ["int2", "int3"]) 
+    # 3 is ranged, 2 has sos
+end
```
Comment (odow) [inline test/MOIWrapper.jl]: Go see the Gurobi and GLPK tests, there are a few missing here.

File: src/cpx_model.jl [hunk 22]
```diff
@@ -52,20 +52,22 @@ end
...
+function get_sense(model::Model) 
+    sense_int = c_api_getobjsen(model)
     if sense_int == 1
-        sense = :Min
-    elseif sense_int == -1
-        sense = :Max
-    else
-        error("CPLEX: problem object or environment does not exist")
+        return :Min
+    else
```
Comment (odow) [inline src/cpx_model.jl]: Why did we drop the check for `sense_int == -1` and the `else` error?

File: src/MOIWrapper.jl [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+    return c_api_getpi(model.inner, place)
+end
+
+function LQOI.get_objective_value(model::Optimizer)
+    return c_api_getobjval(model.inner)
+end
+
+function MOI.free!(m::Optimizer)
+    # close_CPLEX(model.inner)   
+end
```
Comment (odow) [inline src/MOIWrapper.jl]: This method is not needed.

**Comment by IssamT** [comment]
I think we can keep the upgrade to julia 0.7 for another PR, since the MathProgBase interface is also failing it.

File: test/MOIWrapper.jl [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    solver = CPLEX.Optimizer()
+    MOIT.intlineartest(solver, intconfig, ["int2", "int3"]) 
+    # 3 is ranged, 2 has sos
+    @testset "int3" begin
+        MOIT.int3test(
+            MOIB.SplitInterval{Float64}(CPLEX.Optimizer()),
+            MOIT.TestConfig()
+        )
+    end
+end
```
[the remark this answers is not in the document]
Comment (IssamT) [reply test/MOIWrapper.jl]: Done

File: test/MOIWrapper.jl [hunk 25]
```diff
@@ -57,4 +57,24 @@ end
             MOIT.TestConfig()
         )
     end
+end
+
+@testset "ModelLike tests" begin
+    solver = CPLEX.Optimizer()
+    # TODO add a test config to disable testing problem name.
+    # @testset "nametest" begin
+    #     MOIT.nametest(solver)
```
Comment (odow) [inline test/MOIWrapper.jl]: Why does the nametest fail? Since CLPEX uses LQOI it should pass...

Comment (IssamT) [reply test/MOIWrapper.jl]: You're right. My bad.

**Comment by IssamT** [comment]
I think it s better to do the update for MOI 0.6 in a separate PR, to keep a version for comparaison. Any minor issues left can be discussed in the MOI 0.6 update PR.

**Comment by odow** [comment]
Okay, I have tagged a release of CPLEX prior to merging this: https://github.com/JuliaLang/METADATA.jl/pull/17661

I'll squash and merge this, then you can make another PR updating to MOI v0.6 and Julia v0.7. Then we can tag as v0.4.0
